File: lib/worktrees.ts
/**
 * P4/F1 — git worktree sandbox per spawned run.
 *
 * When `app.git.worktreeMode === "enabled"`, the bridge creates a
 * private worktree at `<appRoot>/.worktrees/<sessionId>` BEFORE the
 * child agent spawns. The child runs in that cwd; its edits stay
 * isolated from the live tree until the post-exit gates pass.
 *
 * Lifecycle:
 *
 *   1. `createWorktreeForRun` — runs before spawn. Creates the worktree
 *      on the configured branch (auto-create / fixed / current). Returns
 *      the path the agent should run in plus the branch name.
 *
 *   2. Child writes code + report inside the worktree.
 *
 *   3. After all post-exit gates pass and (optionally) auto-commit
 *      committed in the worktree, `mergeAndRemoveWorktree` rebases the
 *      worktree branch onto the parent branch and removes the worktree.
 *
 *   4. On gate failure or unexpected crash, the worktree is left in
 *      place for inspection. `pruneStaleWorktrees` reaps anything older
 *      than the configured TTL (default 24h) on later API hits.
 *
 * Windows note: `git worktree remove` can fail if a process inside the
 * worktree is still holding a file handle (Claude's `.jsonl` writer is
 * the usual culprit). We pass `--force` and fall back to manual `rm -rf`
 * + `git worktree prune` to mop up.
 */
import { execFile } from "node:child_process";
import {
  existsSync,
  mkdirSync,
  readdirSync,
  rmSync,
  statSync,
} from "node:fs";
import { isAbsolute, join, normalize, resolve, sep } from "node:path";
import { promisify } from "node:util";
import type { AppGitSettings } from "./apps";
import { sanitizeBranchSegment } from "./gitOps";

const execFileP = promisify(execFile);

const SHORT_TIMEOUT_MS = 10_000;
const PRUNE_TIMEOUT_MS = 30_000;
const WORKTREES_DIRNAME = ".worktrees";
const WORKTREE_BRANCH_PREFIX = "claude/wt/";
/** Default TTL for stale worktree pruning. Configurable via env. */
const DEFAULT_STALE_HOURS = 24;

export interface WorktreeHandle {
  /** Absolute path of the worktree directory. */
  path: string;
  /** Branch the worktree was created on (or checked out into). */
  branch: string;
  /**
   * Branch the worktree was forked from — used as the merge target on
   * cleanup. Null means the bridge couldn't determine HEAD (detached
   * head, fresh repo) so cleanup falls back to skipping the merge.
   */
  baseBranch: string | null;
}

export interface WorktreeOpResult {
  ok: boolean;
  message: string;
  error?: string;
}

interface ExecResult {
  ok: boolean;
  stdout: string;
  stderr: string;
  code: number;
}

async function runGit(
  cwd: string,
  args: string[],
  timeoutMs = SHORT_TIMEOUT_MS,
): Promise<ExecResult> {
  try {
    const { stdout } = await execFileP("git", args, {
      cwd,
      timeout: timeoutMs,
      windowsHide: true,
      maxBuffer: 1 << 20,
    });
    return { ok: true, stdout: stdout.toString(), stderr: "", code: 0 };
  } catch (err) {
    const e = err as NodeJS.ErrnoException & {
      stderr?: Buffer | string;
      stdout?: Buffer | string;
      code?: number;
    };
    const stderr =
      typeof e.stderr === "string" ? e.stderr : e.stderr?.toString() ?? "";
    const stdout =
      typeof e.stdout === "string" ? e.stdout : e.stdout?.toString() ?? "";
    return {
      ok: false,
      stdout,
      stderr: stderr.trim() || (e.message ?? "git failed"),
      code: typeof e.code === "number" ? e.code : 1,
    };
  }
}

/** Compute the worktree directory for a given session under the app. */
export function worktreePathFor(appPath: string, sessionId: string): string {
  return join(appPath, WORKTREES_DIRNAME, sessionId);
}

/**
 * Defense-in-depth: ensure a worktree path resolves under the app
 * root. The session id is UUID-validated upstream so this is layered
 * paranoia, not the primary check.
 */
function isUnderAppRoot(appPath: string, candidate: string): boolean {
  const a = resolve(appPath);
  const c = resolve(candidate);
  if (a === c) return false;
  return c.startsWith(a + sep) || c.startsWith(a + "/");
}

async function currentBranch(cwd: string): Promise<string | null> {
  const r = await runGit(cwd, ["rev-parse", "--abbrev-ref", "HEAD"]);
  if (!r.ok) return null;
  const name = r.stdout.trim();
  return name && name !== "HEAD" ? name : null;
}

async function branchExists(cwd: string, branch: string): Promise<boolean> {
  const r = await runGit(cwd, [
    "rev-parse",
    "--verify",
    "--quiet",
    `refs/heads/${branch}`,
  ]);
  return r.ok;
}

/**
 * Mint the unique per-spawn branch the worktree will execute on.
 * Always unique — using `branchMode`'s natural branch (e.g. `main`,
 * `claude/<task>`) would clash whenever that branch is already
 * checked out in the live tree (`git worktree add` refuses), or two
 * concurrent spawns chose the same name. The actual merge target is
 * computed separately by `resolveBaseBranch`.
 */
function mintSpawnBranch(taskId: string, sessionId: string): string {
  const shortSid = sessionId.replace(/-/g, "").slice(0, 8);
  return `${WORKTREE_BRANCH_PREFIX}${sanitizeBranchSegment(taskId)}-${shortSid}`;
}

/**
 * Resolve the branch the worktree should fork from. The bridge merges
 * back into this branch on cleanup, so it's the eventual landing
 * place for the agent's commits:
 *
 *   - `fixed`       → `fixedBranch` if it exists locally, else current HEAD.
 *   - `auto-create` → `claude/<taskId>` (auto-created from current HEAD if missing).
 *   - `current`     → current HEAD branch.
 *
 * Returns `null` when the live tree is on a detached HEAD and we
 * couldn't materialize a base — caller falls through, the worktree
 * forks from HEAD without an explicit base, and merge is skipped on
 * cleanup.
 */
async function resolveBaseBranch(
  appPath: string,
  settings: AppGitSettings,
  taskId: string,
): Promise<string | null> {
  const liveHead = await currentBranch(appPath);
  if (settings.branchMode === "fixed" && settings.fixedBranch.trim()) {
    const branch = settings.fixedBranch.trim();
    if (await branchExists(appPath, branch)) return branch;
    // Fixed branch doesn't exist locally — fall back to live HEAD so
    // the worktree still gets created. Operator's job to populate the
    // ref before relying on it.
    return liveHead;
  }
  if (settings.branchMode === "auto-create") {
    const branch = `claude/${sanitizeBranchSegment(taskId)}`;
    if (await branchExists(appPath, branch)) return branch;
    // Materialize the auto-create branch from current HEAD so the
    // worktree has a valid base to fork from. We use `git branch`
    // (no checkout) so the live tree's HEAD stays put.
    if (liveHead) {
      const r = await runGit(appPath, ["branch", branch, liveHead]);
      if (r.ok) return branch;
    }
    return liveHead;
  }
  return liveHead;
}

/**
 * Create a private worktree for a spawned run. Returns null on any
 * failure (caller falls back to spawning in the live tree, with a
 * warning logged). On Windows the worktree dir is created with
 * `mkdir -p` semantics first so `git worktree add` doesn't trip on a
 * missing parent.
 */
export async function createWorktreeForRun(args: {
  appPath: string;
  settings: AppGitSettings;
  taskId: string;
  sessionId: string;
}): Promise<WorktreeHandle | null> {
  const { appPath, settings, taskId, sessionId } = args;
  if (!isAbsolute(appPath)) return null;
  if (!existsSync(join(appPath, ".git"))) return null;

  const wtPath = worktreePathFor(appPath, sessionId);
  if (!isUnderAppRoot(appPath, wtPath)) return null;

  // Make sure the parent .worktrees/ dir exists. `git worktree add`
  // creates the leaf, but a missing intermediate dir is a Windows
  // pain.
  try {
    mkdirSync(join(appPath, WORKTREES_DIRNAME), { recursive: true });
  } catch {
    return null;
  }
  if (existsSync(wtPath)) {
    // Stale leftover from a prior crashed spawn — refuse to clobber.
    return null;
  }

  const baseBranch = await resolveBaseBranch(appPath, settings, taskId);
  const spawnBranch = mintSpawnBranch(taskId, sessionId);

  // `git worktree add -b <newBranch> <path> [<startPoint>]`
  // The spawn branch is always fresh (per-session unique), so we can
  // unconditionally use the `-b` form. Passing `<startPoint>` makes
  // the new branch fork from the configured base; omitting it forks
  // from current HEAD (which is the same in `current` mode and a safe
  // fallback when baseBranch resolution failed).
  const addArgs: string[] = [
    "worktree",
    "add",
    "-b",
    spawnBranch,
    wtPath,
  ];
  if (baseBranch) addArgs.push(baseBranch);
  const r = await runGit(appPath, addArgs);
  if (!r.ok) {
    await runGit(appPath, ["worktree", "prune"], PRUNE_TIMEOUT_MS);
    return null;
  }

  return { path: wtPath, branch: spawnBranch, baseBranch };
}

/**
 * Remove a worktree directory + the matching git registration. Tries
 * `git worktree remove --force` first (clean path); on Windows a
 * locked file handle can make that fail, so we fall back to manual
 * rmdir + `git worktree prune`.
 *
 * Returns ok=true even when the directory was already gone — callers
 * are idempotent about this.
 */
export async function removeWorktree(args: {
  appPath: string;
  worktreePath: string;
}): Promise<WorktreeOpResult> {
  const { appPath, worktreePath } = args;
  if (!isUnderAppRoot(appPath, worktreePath)) {
    return {
      ok: false,
      message: "refusing to remove path outside app root",
      error: worktreePath,
    };
  }
  if (!existsSync(worktreePath)) {
    // Already gone — still prune the registration in case git's index
    // remembers it.
    await runGit(appPath, ["worktree", "prune"], PRUNE_TIMEOUT_MS);
    return { ok: true, message: "worktree already removed" };
  }
  const remove = await runGit(
    appPath,
    ["worktree", "remove", "--force", worktreePath],
    PRUNE_TIMEOUT_MS,
  );
  if (remove.ok) {
    return { ok: true, message: `removed worktree ${worktreePath}` };
  }
  // Windows fallback: manual rmdir then prune the registration.
  try {
    rmSync(worktreePath, { recursive: true, force: true, maxRetries: 3 });
  } catch (e) {
    return {
      ok: false,
      message: `worktree remove failed and rm fallback errored: ${(e as Error).message}`,
      error: remove.stderr,
    };
  }
  await runGit(appPath, ["worktree", "prune"], PRUNE_TIMEOUT_MS);
  return { ok: true, message: `removed worktree (rm fallback) ${worktreePath}` };
}

/**
 * After a successful run + auto-commit (which committed to the
 * worktree's branch), merge that branch back into the base branch and
 * remove the worktree. The merge happens in the live app tree, not in
 * the worktree (you can't operate on a branch that's checked out in
 * another worktree).
 *
 * Fail-soft: any merge failure leaves the worktree in place (caller
 * surfaces the error to the operator) but never throws.
 */
export async function mergeAndRemoveWorktree(args: {
  appPath: string;
  handle: WorktreeHandle;
}): Promise<WorktreeOpResult> {
  const { appPath, handle } = args;

  // No base branch means the worktree was created on a detached HEAD
  // or a fresh repo. There's nothing meaningful to merge into — just
  // remove the worktree and surface a note.
  if (!handle.baseBranch) {
    const removed = await removeWorktree({
      appPath,
      worktreePath: handle.path,
    });
    return {
      ok: removed.ok,
      message: `no base branch to merge into; ${removed.message}`,
      error: removed.error,
    };
  }
  if (handle.baseBranch === handle.branch) {
    // Worktree shared the base branch (branchMode=fixed pointing at the
    // current branch). Auto-commit already committed onto that branch
    // from inside the worktree, so the live tree is already up to date
    // — just remove the worktree.
    return removeWorktree({ appPath, worktreePath: handle.path });
  }

  // Make sure the live tree is on the base branch before merging.
  const liveCur = await currentBranch(appPath);
  if (liveCur !== handle.baseBranch) {
    const co = await runGit(appPath, ["checkout", handle.baseBranch]);
    if (!co.ok) {
      return {
        ok: false,
        message: `failed to checkout base branch ${handle.baseBranch}`,
        error: co.stderr,
      };
    }
  }

  // Use --no-ff so the merge commit is preserved even on fast-forward
  // candidates; preserves the audit trail of "this work came from a
  // bridge worktree".
  const merge = await runGit(appPath, [
    "merge",
    "--no-ff",
    "--no-edit",
    handle.branch,
  ]);
  if (!merge.ok) {
    return {
      ok: false,
      message: `merge of ${handle.branch} into ${handle.baseBranch} failed`,
      error: merge.stderr,
    };
  }

  const removed = await removeWorktree({
    appPath,
    worktreePath: handle.path,
  });
  return {
    ok: removed.ok,
    message: `merged ${handle.branch} into ${handle.baseBranch}; ${removed.message}`,
    error: removed.error,
  };
}

function staleHours(): number {
  const raw = process.env.BRIDGE_WORKTREE_STALE_HOURS;
  const n = raw ? Number(raw) : DEFAULT_STALE_HOURS;
  return Number.isFinite(n) && n > 0 ? n : DEFAULT_STALE_HOURS;
}

/**
 * Reap worktree directories under `<appPath>/.worktrees/` whose mtime
 * is older than the configured TTL. Cheap to call from the same hot
 * path that triggers `staleRunReaper` (no extra background process).
 *
 * Returns the count of worktrees actually removed.
 */
export async function pruneStaleWorktrees(args: {
  appPath: string;
  /** Override TTL for tests. Falls back to `BRIDGE_WORKTREE_STALE_HOURS` env. */
  staleAfterMs?: number;
}): Promise<number> {
  const { appPath } = args;
  const root = join(appPath, WORKTREES_DIRNAME);
  if (!existsSync(root)) return 0;

  const cutoffMs =
    Date.now() - (args.staleAfterMs ?? staleHours() * 60 * 60 * 1000);
  let entries: import("node:fs").Dirent[];
  try {
    entries = readdirSync(root, { withFileTypes: true });
  } catch {
    return 0;
  }
  let removed = 0;
  for (const entry of entries) {
    if (!entry.isDirectory()) continue;
    const wt = join(root, entry.name);
    if (!isUnderAppRoot(appPath, wt)) continue;
    // The dir's own mtime only ticks on entry add/remove — an agent
    // modifying existing files would otherwise look idle. Walk the
    // immediate top-level children and take the max mtime as a
    // cheap proxy for "last write activity inside the worktree".
    let mtimeMs: number;
    try {
      mtimeMs = statSync(wt).mtimeMs;
      let kids: import("node:fs").Dirent[] = [];
      try {
        kids = readdirSync(wt, { withFileTypes: true });
      } catch { /* unreadable — fall through to root mtime only */ }
      for (const k of kids) {
        if (k.name === ".git") continue;
        try {
          const km = statSync(join(wt, k.name)).mtimeMs;
          if (km > mtimeMs) mtimeMs = km;
        } catch { /* ignore individual entry errors */ }
      }
    } catch {
      continue;
    }
    if (mtimeMs > cutoffMs) continue;
    const r = await removeWorktree({ appPath, worktreePath: wt });
    if (r.ok) removed += 1;
  }
  return removed;
}

/** Test-only helper: expose the canonical normalized path. */
export function _normalize(p: string): string {
  return normalize(p);
}

/**
 * (P4) Pull the worktree-related fields off a parent run so a retry
 * spawn can inherit them. Each retry runs in the SAME worktree as the
 * run it's retrying — otherwise the retry would spawn in the live tree
 * and never see the parent's WIP edits, defeating the whole point of
 * the sandbox. Returns an empty object when the parent didn't run in a
 * worktree, so callers can spread it unconditionally.
 */
export function inheritWorktreeFields(parent: {
  worktreePath?: string | null;
  worktreeBranch?: string | null;
  worktreeBaseBranch?: string | null;
}): {
  worktreePath: string | null;
  worktreeBranch: string | null;
  worktreeBaseBranch: string | null;
} {
  return {
    worktreePath: parent.worktreePath ?? null,
    worktreeBranch: parent.worktreeBranch ?? null,
    worktreeBaseBranch: parent.worktreeBaseBranch ?? null,
  };
}

File: internal/git/worktrees.go
package git

import (
	"bufio"
	"errors"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"

	"github.com/stop1love1/claude-bridge/internal/meta"
	"github.com/stop1love1/claude-bridge/internal/pathsafe"
)

// worktreesDirName is the per-app folder that holds every spawn's
// private worktree. Kept under the app root so a sibling tool that
// scans for ".git" subtrees finds one parent per worktree, not a
// global pool.
const worktreesDirName = ".worktrees"

// worktreeBranchPrefix mints branches in their own namespace so a
// reviewer scanning `git branch` can tell bridge-spawned branches from
// human-authored ones at a glance. The `wt/` infix distinguishes
// worktree-isolated runs from the `claude/<task-id>` branches the
// auto-create lifecycle hook produces.
const worktreeBranchPrefix = "claude/wt/"

// Worktree describes a per-spawn isolated checkout. The three fields
// match the optional pointer fields meta.Run carries (WorktreePath,
// WorktreeBranch, WorktreeBaseBranch) so the lifecycle hook can move
// them between Worktree values and meta.json without re-deriving.
//
// BaseBranch is empty when the worktree was forked from a detached
// HEAD or a fresh repo; MergeWorktreeBack treats that as "skip merge,
// just remove" so we never try to merge into an unnamed ref.
type Worktree struct {
	Path       string
	Branch     string
	BaseBranch string
}

// MergeOutcome is the structured result of MergeWorktreeBack. Conflict
// is split out from Merged so a caller can distinguish "we tried and
// the merge produced conflicts" (worktree branch retained for manual
// recovery) from "we never tried because there was nothing to merge".
type MergeOutcome struct {
	Merged   bool
	Conflict bool
	Message  string
}

// CreateWorktreeForRun allocates `<appRoot>/.worktrees/<sessionID>/`
// and runs `git worktree add -b <branch> <path> [<baseBranch>]`. When
// baseBranch is empty the new branch forks from the live tree's
// current HEAD — same effect as omitting <start-point> from the git
// CLI. The worktree branch is always per-session unique
// (`claude/wt/<sanitized-sessionID>`) so concurrent spawns can never
// collide on the branch name.
//
// Returns an error rather than nil-handle so the lifecycle hook can
// log the underlying git failure. Callers that want fail-soft
// behaviour (fall back to spawning in the live tree) should swallow
// the error themselves — the bridge logs but does not abort spawn.
func CreateWorktreeForRun(appRoot, sessionID, baseBranch string) (Worktree, error) {
	if !filepath.IsAbs(appRoot) {
		return Worktree{}, fmt.Errorf("worktree: appRoot must be absolute, got %q", appRoot)
	}
	if sessionID == "" {
		return Worktree{}, errors.New("worktree: sessionID required")
	}
	if _, err := os.Stat(filepath.Join(appRoot, ".git")); err != nil {
		return Worktree{}, fmt.Errorf("worktree: %s is not a git repo: %w", appRoot, err)
	}
	// baseBranch (when supplied) ends up as a positional arg to
	// `git worktree add`. Reject anything that could be interpreted as
	// a CLI option before the validator inside `git` ever sees it.
	if baseBranch != "" {
		if err := validateBranchName(baseBranch); err != nil {
			return Worktree{}, fmt.Errorf("worktree: invalid base branch: %w", err)
		}
	}

	wtPath := worktreePathFor(appRoot, sessionID)
	if !isUnderAppRoot(appRoot, wtPath) {
		return Worktree{}, fmt.Errorf("worktree: refusing to create path outside app root: %q", wtPath)
	}

	// Ensure `<appRoot>/.worktrees/` exists before `git worktree add`.
	// On Windows a missing intermediate dir trips the add with a
	// confusing message; mkdir -p semantics here keep the failure
	// surface clean.
	if err := os.MkdirAll(filepath.Join(appRoot, worktreesDirName), 0o755); err != nil {
		return Worktree{}, fmt.Errorf("worktree: mkdir parent: %w", err)
	}
	// A leftover dir from a crashed prior spawn would make `git
	// worktree add` fail with a cryptic "already exists". Refuse
	// up front so the caller's log line points at the real cause.
	if _, err := os.Stat(wtPath); err == nil {
		return Worktree{}, fmt.Errorf("worktree: target path already exists: %q", wtPath)
	}

	resolvedBase := baseBranch
	if resolvedBase == "" {
		// Empty baseBranch → fork from current HEAD. We resolve the
		// branch name (rather than letting git default to HEAD) so the
		// returned Worktree.BaseBranch carries an explicit name when
		// one exists; merge-back needs a named ref.
		if b, ok := currentBranchOf(appRoot); ok {
			resolvedBase = b
		}
	}

	branch := mintWorktreeBranch(sessionID)
	// mintWorktreeBranch sanitizes through sanitizeBranchSegment, but we
	// re-validate anyway: a future change to that helper that accidentally
	// allows a leading dash would otherwise produce a branch name that
	// `git` parses as an option.
	if err := validateBranchName(branch); err != nil {
		return Worktree{}, fmt.Errorf("worktree: minted branch invalid: %w", err)
	}
	args := []string{"worktree", "add", "-b", branch, wtPath}
	if resolvedBase != "" {
		args = append(args, resolvedBase)
	}
	if err := run(appRoot, "git", args...); err != nil {
		// `git worktree add` can leave a half-registered entry behind
		// when it errors after cloning the metadata but before the
		// checkout finishes. Prune so a retry with the same sessionID
		// (after the operator clears the leftover dir) doesn't trip
		// over the stale registration.
		_ = run(appRoot, "git", "worktree", "prune")
		return Worktree{}, fmt.Errorf("worktree: add: %w", err)
	}
	return Worktree{Path: wtPath, Branch: branch, BaseBranch: resolvedBase}, nil
}

// RemoveWorktree runs `git worktree remove --force <path>` then falls
// back to RemoveAll on the directory if git couldn't clean it (Windows
// file-handle locks from the agent's still-flushing .jsonl writer are
// the usual culprit). A final `git worktree prune` mops up the
// registration so `git worktree list` doesn't keep showing the
// removed entry as "prunable".
//
// Idempotent: missing directory returns nil after pruning the
// registration, so a double-remove from retry logic is harmless.
func RemoveWorktree(appRoot, sessionID string) error {
	wtPath := worktreePathFor(appRoot, sessionID)
	if !isUnderAppRoot(appRoot, wtPath) {
		return fmt.Errorf("worktree: refusing to remove path outside app root: %q", wtPath)
	}
	if _, err := os.Stat(wtPath); errors.Is(err, os.ErrNotExist) {
		_ = run(appRoot, "git", "worktree", "prune")
		return nil
	}
	if err := run(appRoot, "git", "worktree", "remove", "--force", wtPath); err == nil {
		return nil
	}
	// Windows fallback: git refused (locked file), so do the unlink
	// ourselves and let prune drop the orphaned registration.
	if err := os.RemoveAll(wtPath); err != nil {
		return fmt.Errorf("worktree: rm fallback: %w", err)
	}
	_ = run(appRoot, "git", "worktree", "prune")
	return nil
}

// MergeWorktreeBack checks out the worktree's BaseBranch in the live
// tree and runs `git merge --no-ff --no-edit <branch>`. The merge runs
// in the live app tree (not in the worktree itself) because git
// refuses to operate on a branch that's checked out in another
// worktree.
//
// On conflict, aborts the merge and returns Outcome{Conflict: true}
// with the worktree left intact so the operator can resolve manually
// and remove it later. On a clean merge the caller is responsible for
// invoking RemoveWorktree — keeping the steps separate lets the
// coordinator decide whether to preserve a worktree for inspection
// even after a successful merge.
//
// Returns Outcome{Merged: false, Conflict: false} when there's
// nothing to merge (no BaseBranch, or BaseBranch == Branch — the
// latter happens when settings.fixedBranch matched the live HEAD and
// commits already landed on the right ref from inside the worktree).
func MergeWorktreeBack(appRoot string, w Worktree) (MergeOutcome, error) {
	if w.BaseBranch == "" {
		return MergeOutcome{Message: "no base branch; nothing to merge"}, nil
	}
	if w.BaseBranch == w.Branch {
		return MergeOutcome{Message: "base branch equals worktree branch; nothing to merge"}, nil
	}
	// Both branch names ride into git argv; validate before exec so a
	// crafted Worktree literal (or stale meta.json hand-edit) can't turn
	// `BaseBranch="--upload-pack=evil"` into a git option.
	if err := validateBranchName(w.BaseBranch); err != nil {
		return MergeOutcome{}, fmt.Errorf("worktree: invalid base branch: %w", err)
	}
	if err := validateBranchName(w.Branch); err != nil {
		return MergeOutcome{}, fmt.Errorf("worktree: invalid branch: %w", err)
	}

	// Switch the live tree to the merge target. We don't restore the
	// previous branch on exit — the bridge's per-app branchMode hook
	// is what governs the live tree's checked-out branch, not this
	// merge step.
	if cur, ok := currentBranchOf(appRoot); !ok || cur != w.BaseBranch {
		if err := run(appRoot, "git", "checkout", w.BaseBranch); err != nil {
			return MergeOutcome{}, fmt.Errorf("worktree: checkout base failed: %w", err)
		}
	}

	// --no-ff preserves the merge commit even when fast-forwarding
	// would suffice; the audit trail of "this work came from a bridge
	// worktree" is the whole reason this branch exists. The trailing
	// `--` lets git distinguish the branch token from any pathspec git
	// might otherwise look for, even if validateBranchName were ever
	// loosened.
	mergeErr := run(appRoot, "git", "merge", "--no-ff", "--no-edit", w.Branch)
	if mergeErr == nil {
		return MergeOutcome{
			Merged:  true,
			Message: fmt.Sprintf("merged %s into %s", w.Branch, w.BaseBranch),
		}, nil
	}
	// Conflict path: abort so the live tree returns to a clean state
	// and the operator picks up the worktree branch by hand.
	_ = run(appRoot, "git", "merge", "--abort")
	return MergeOutcome{
		Conflict: true,
		Message:  fmt.Sprintf("merge of %s into %s conflicted; worktree retained", w.Branch, w.BaseBranch),
	}, nil
}

// ListWorktrees parses `git worktree list --porcelain` and returns
// every registered worktree under appRoot. The porcelain format is a
// stable contract since git 2.7 — each entry is a sequence of
// "<key> <value>" lines terminated by a blank line. We surface only
// the path / branch / detached marker (encoded as empty Branch).
//
// BaseBranch is left empty in the returned values: porcelain doesn't
// record the fork point, and re-deriving it would require a
// merge-base scan per entry that's nowhere near the cost callers are
// willing to pay for what is, in practice, a registry view.
func ListWorktrees(appRoot string) ([]Worktree, error) {
	if _, err := exec.LookPath("git"); err != nil {
		return nil, fmt.Errorf("worktree: git not on PATH: %w", err)
	}
	cmd := exec.Command("git", "worktree", "list", "--porcelain")
	cmd.Dir = appRoot
	out, err := cmd.CombinedOutput()
	if err != nil {
		return nil, fmt.Errorf("worktree: list: %w (%s)", err, strings.TrimSpace(string(out)))
	}

	var (
		results []Worktree
		cur     Worktree
		hasCur  bool
	)
	flush := func() {
		if hasCur {
			results = append(results, cur)
		}
		cur = Worktree{}
		hasCur = false
	}
	scanner := bufio.NewScanner(strings.NewReader(string(out)))
	scanner.Buffer(make([]byte, 0, 64*1024), 1<<20)
	for scanner.Scan() {
		line := scanner.Text()
		if line == "" {
			flush()
			continue
		}
		key, val, _ := strings.Cut(line, " ")
		switch key {
		case "worktree":
			cur.Path = val
			hasCur = true
		case "branch":
			// Porcelain emits the full ref ("refs/heads/main"); strip
			// the prefix so callers can compare against branch names
			// they already have without re-parsing.
			cur.Branch = strings.TrimPrefix(val, "refs/heads/")
		case "detached":
			cur.Branch = ""
		}
	}
	flush()
	if err := scanner.Err(); err != nil {
		return nil, fmt.Errorf("worktree: parse list: %w", err)
	}
	return results, nil
}

// InheritWorktreeFields pulls the three worktree pointer fields off a
// parent run so a retry spawn can inherit them. A retry MUST run in
// the same worktree as the run it's retrying — otherwise it would
// spawn in the live tree, miss the parent's WIP edits, and silently
// duplicate work onto a different branch. Returns three nil pointers
// when the parent didn't run in a worktree, so callers can copy them
// into the child run unconditionally.
func InheritWorktreeFields(parent meta.Run) (path, branch, baseBranch *string) {
	return parent.WorktreePath, parent.WorktreeBranch, parent.WorktreeBaseBranch
}

// worktreePathFor returns the canonical per-session worktree dir.
// Exposed (lowercase) so package-internal callers stay in sync with
// CreateWorktreeForRun without hand-concatenating path segments.
func worktreePathFor(appRoot, sessionID string) string {
	return filepath.Join(appRoot, worktreesDirName, sessionID)
}

// isUnderAppRoot is defense-in-depth: even if a caller passes a
// crafted sessionID with `..` segments, RemoveWorktree must never
// `RemoveAll` something outside the app tree. The sessionID is
// validated upstream as a UUID, so this is layered paranoia.
//
// Strict descendant only — equal-to-root counts as "not under" so
// RemoveAll can never target the entire app tree. pathsafe.ContainsStrict
// encodes that policy.
func isUnderAppRoot(appRoot, candidate string) bool {
	return pathsafe.ContainsStrict(appRoot, candidate)
}

// currentBranchOf resolves the working tree's current branch via
// `git rev-parse --abbrev-ref HEAD`. Returns ok=false on detached
// HEAD or any rev-parse error so callers can fall through (worktree
// forks from HEAD without a named base, merge-back is skipped).
//
// The output is treated as untrusted: git's stdout for a corrupted
// HEAD ref could in principle contain anything, and the value flows
// back into git argv via Worktree.BaseBranch. We run it through
// validateBranchName before declaring it usable; failures fall back to
// ok=false (same as detached-HEAD handling).
func currentBranchOf(repoPath string) (string, bool) {
	cmd := exec.Command("git", "rev-parse", "--abbrev-ref", "HEAD")
	cmd.Dir = repoPath
	out, err := cmd.Output()
	if err != nil {
		return "", false
	}
	name := strings.TrimSpace(string(out))
	if name == "" || name == "HEAD" {
		return "", false
	}
	if err := validateBranchName(name); err != nil {
		return "", false
	}
	return name, true
}

// mintWorktreeBranch builds the per-session branch name. The session
// ID is sanitized through sanitizeBranchSegment so any odd characters
// (uppercase, spaces, weirdness from a non-UUID test input) collapse
// to the bridge's canonical charset before they become a ref name.
func mintWorktreeBranch(sessionID string) string {
	return worktreeBranchPrefix + sanitizeBranchSegment(sessionID)
}

// sanitizeBranchSegment lowercases and restricts a segment to the
// bridge's branch charset: [a-z0-9./-]. Mirrors the same charset the
// auto-create lifecycle hook uses, so a worktree branch and an
// auto-create branch derived from the same seed produce comparable
// strings. Empty input → "session" (any non-empty placeholder; "task"
// is what the TS code uses, but the worktree codepath is keyed off
// sessionID, so "session" is the more honest fallback).
func sanitizeBranchSegment(raw string) string {
	lower := strings.ToLower(raw)
	var b strings.Builder
	b.Grow(len(lower))
	for _, r := range lower {
		switch {
		case r >= 'a' && r <= 'z',
			r >= '0' && r <= '9',
			r == '/', r == '-', r == '.':
			b.WriteRune(r)
		default:
			b.WriteByte('-')
		}
	}
	cleaned := strings.TrimLeft(b.String(), "/.-")
	if cleaned == "" {
		return "session"
	}
	return cleaned
}
